### PR TITLE
Use fused types for _take_2d

### PR DESCRIPTION
**Pull request by jbrockmendel**
Misc cleanup

For some of the cython code I get unexplained test failures when using fused types.  Vaguely hoping to narrow down the set of affected functions so as to make a helpful bug report to cython.

### CODE REVIEW

File: pandas/_libs/algos_take_helper.pxi.in [hunk 1]
```diff
@@ -260,33 +260,39 @@ def take_2d_multi_{{name}}_{{dest}}(ndarray[{{c_type_in}}, ndim=2] values,
...
-        ndarray[{{ctype}}, ndim=2] result
+        ndarray[take_t, ndim=2] result
         object val
 
     N, K = (<object> values).shape
-    result = {{init_result}}
+
+    if take_t is object:
+        # evaluated at compile-time
+        result = values.copy()
```
Comment (jreback) [inline pandas/_libs/algos_take_helper.pxi.in]: empty_like doesn't work?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/algos_take_helper.pxi.in (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/algos_take_helper.pxi.in]: This is just maintaining the status quo, so I didn't try it the other way.

**Comment by jreback** [comment]
rebase. I assume you want to merge this, or was this just testing things?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**